File: rules-runner/rules-runner-kotlin/src/main/kotlin/eu/ehn/dcc/rulesets/typings.kt
package eu.ehn.dcc.rulesets

import com.fasterxml.jackson.databind.JsonNode


data class Rule(
    var id: String,
    var description: String,
    var logic: JsonNode
)
typealias RuleSet = List<Rule>

sealed class RuleEvaluationResult {
    data class BooleanResult(
        val value: Boolean
    ) : RuleEvaluationResult()
    data class EvaluationError(
        val errorMessage: String
    ) : RuleEvaluationResult()
}

typealias RuleEvaluations = Map<String, RuleEvaluationResult>

data class RuleSetEvaluationResult(
    val ruleEvaluations: RuleEvaluations,
    val hasErrors: Boolean,
    val allSatisfied: Boolean
)

File: rules-runner/javascript/rules-runner-js/src/typings.ts
import { CertLogicExpression } from "certlogic-js"


/**
 * Types representing rules and rule sets, as read in from a JSON file.
 */
export type Rule = {
    id: string
    logic: CertLogicExpression
}
export type RuleSet = Rule[]


/**
 * A type representing the “compressed” value sets.
 */
export type ValueSets = { [valueSetId: string]: string[] }


/**
 * A type representing the external parameters object.
 */
export type ExternalParameters = {
    valueSets: ValueSets
    [key: string]: unknown  // Allow any other property.
}

/**
 * A type representing the data context passed to every rule.
 */
export type RuleEvaluationDataContext = {
    payload: any,
    external: ExternalParameters
}


/**
 * A type representing the evaluation of a single rule.
 */
export type RuleEvaluationResult = boolean | { errorMessage: string }

/**
 * A type representing the evaluation of all rules (individually) in a rule set.
 */
export type RuleEvaluations = { [ruleId: string]: RuleEvaluationResult }

/**
 * A type representing the evaluation of all rules in a rule set,
 * with derived properties for whether errors occurred,
 * and whether all rules passed.
 */
export type RuleSetEvaluationResult = {
    ruleEvaluations: RuleEvaluations
    hasErrors: boolean
    allSatisfied: boolean
}

